### PR TITLE
Bugfix/bug 864259

### PR DESCRIPTION
**Pull request by palao**
This should close https://bugs.gentoo.org/864259

### CODE REVIEW

**Comment by palao** [comment]
AFAICS, the failures have nothing to do with this PR, am I wrong?

**Comment by floppym** [comment]
> AFAICS, the failures have nothing to do with this PR, am I wrong?

I'm not sure what gives you that impression; it looks like this PR does cause the failure. It will need to be investigated before this can be merged.

**Comment by palao** [comment]
It's true. My bad, sorry. I'll have a look at it.
Thank you for the comment.

**Comment by palao** [comment]
I fixed the problem following a different approach (since the other was broken).
As I'm quite new to portage I would appreciate a critical view.

Among all possible solutions that I saw, I tried the one that is more testable and less invasive. I hope I did it well...

**Comment by floppym** [comment]
I plan to take a look at this tomorrow.

**Comment by floppym** [comment]
Sorry, but I have to ask you to clean up the commit history a bit.

- Rebase your changes on current master (no merge commits!).
- Some of the commits undo changes made in prior commits. Please squash these into a single commit.
- In cases where you are making large amounts of whitespace changes (de-indenting blocks), please move the whitespace changes into a separate commit. We can then add that commit to .git-blame-ignore-revs.

**Comment by palao** [comment]
Thank you for the feedback.
I'll hope to do the git surgery by Friday. But I cannot promise: these days I'm quite busy.

**Comment by palao** [comment]
> Sorry, but I have to ask you to clean up the commit history a bit.
> 
>     * Rebase your changes on current master (no merge commits!).
> 
>     * Some of the commits undo changes made in prior commits. Please squash these into a single commit.
> 
>     * In cases where you are making large amounts of whitespace changes (de-indenting blocks), please move the whitespace changes into a separate commit. We can then add that commit to .git-blame-ignore-revs.

Done it. Please let me know if I missed something, or there is another problem.

**Comment by palao** [comment]
Hmmm. Before cleaning up the history that was passing. During the git surgery I didn't touch anything related to the (what I see is) the fail.
Any thoughts?

**Comment by floppym** [comment]
> Hmmm. Before cleaning up the history that was passing. During the git surgery I didn't touch anything related to the (what I see is) the fail.
> Any thoughts?

Don't worry about it: there's a flakey test that passed after I restarted it. See bug [850127](https://bugs.gentoo.org/850127).

**Comment by thesamesam** [comment]
@syu-nya

**Comment by syu-nya** [comment]
I would say set `getbinpkg_refresh=False` at https://github.com/gentoo/portage/blob/06e43403a8d198cbf67c4dab81c799351f44f02f/lib/_emerge/actions.py#L171-L174
should be enough.

There are only 3 places in portage will call `populate()` with binpkg index.

The db init at 
https://github.com/gentoo/portage/blob/06e43403a8d198cbf67c4dab81c799351f44f02f/lib/_emerge/actions.py#L3512-L3514

this issue, which will always called after db init 
https://github.com/gentoo/portage/blob/06e43403a8d198cbf67c4dab81c799351f44f02f/lib/_emerge/actions.py#L172-L174

and followed
https://github.com/gentoo/portage/blob/06e43403a8d198cbf67c4dab81c799351f44f02f/lib/_emerge/actions.py#L429-L431
This one already disabled refresh binpkg index.

I was thought `action_build()` could be called by others but did not found one.

If external program call `action_build()`, maybe 
`getbinpkg_refresh=(not emerge_config.target_config.trees["bintree"].populated)`

File: lib/portage/tests/dbapi/test_bintree.py
Comment (floppym) [inline lib/portage/tests/dbapi/test_bintree.py:109]: I'm confused by this. Why are you passing a string in the getbinpkg_refresh kwarg?

**Comment by floppym** [comment]
I don't have a strong opinion on changing the default versus changing the single call site as @syu-nya suggests.

**Comment by palao** [comment]
@syu-nya, the advantage of your proposal is obviously that it requires to change less code.

However, the idea behind my proposal is as follows. The bug fix must ensure that the remote package indices are transferred only once. If that must be the case, I thought it would be better to be explicit about it: only one call to `binarytree.populate` is explicitly done with `getbinpkg_refresh=True` in the PR, as it is now.

Of course I'm more inclined towards my proposal, but yours is also good. I can change to yours, if necessary.

In any case, all this is telling me that, in my opinion, a deeper change is needed in the interface of the `binarytree` class, particularly the way the `binarytree.populate` method works and must be used. Somehow is not clean. That is clear to me. But, honestly, I don't have a clear proposal to do yet... Still I thought that changing the default behavior to make more explicit what is going on is an improvement.

**Comment by syu-nya** [comment]
@palao I don't have strong prefer, but explicitly to refresh index is a good reason. Indeed a lot of old logic could be improved, but it would be very time consuming.

File: lib/portage/tests/dbapi/test_bintree.py
Comment (palao) [reply lib/portage/tests/dbapi/test_bintree.py:109]: The idea of this test is to make sure that if `populate` is called with `getbinpkgs=True`, then `_populate_remote` is called, and it is *blindly* given whatever is passed in the `getbinpkg_refresh` argument of `populate`.
This test is also telling us that there is no validation done by `populate` and the semantics of the `getbinpkg_refresh` parameter (and all the responsibility to handle it) is translated to `_populate_remote`.

I chose a meaningless string to make clear the point that for `populate` it is irrelevant what is the value of that parameter.

Do you think that this is wrong or unnecessary? Do you have a better idea?

Comment (floppym) [reply lib/portage/tests/dbapi/test_bintree.py:109]: Ah, thanks for explaining.

**Review by floppym — APPROVED** [review]
This PR looks ok to me now. I'll merge tomorrow if there is no negative feedback before then.

**Pull request closed without merging**
